python/taichi/profiler/profiler_impl.py

```diff
@@ -0,0 +1,166 @@
+    '''
+
+    trace = 'trace'
+    count = 'count'
+
+    def __init__(self):
+        self._profiling_mode = False
+        self._total_time_ms = 0.0
+        self._traced_records = []
+        self._statistical_results = {}
+
+    def set_kernel_profiler_mode(self, kernel_profiler=False):
+        if type(kernel_profiler) is bool:
+            self._profiling_mode = kernel_profiler
+        else:
+            raise TypeError(f'Arg kernel_profiler must be of type boolean. '
+                            f'Type {type(kernel_profiler)} '
+                            f'is not supported')
+
+    def get_kernel_profiler_mode(self):
+        return self._profiling_mode
+
+    def clear_fronted(self):
+        self._total_time_ms = 0.0
+        self._traced_records.clear()
+        self._statistical_results.clear()
+
+    def get_total_time(self):
+        self.update_records()  # traced records
+        self.count_results()  # _total_time_ms is counted here
+        return self._total_time_ms
+
+    def query_info(self, name):
+        self.update_records()  # traced records
+        self.count_results()  # statistical results
+        # TODO : query self.StatisticalResult in python scope
+        return impl.get_runtime().prog.query_kernel_profile_info(name)
+
+    def clear_info(self):
+        impl.get_runtime().sync()  #sync first
+        #clear backend & frontend
+        impl.get_runtime().prog.clear_kernel_profile_info()
+        self.clear_fronted()
+
+    def update_records(self):
+        impl.get_runtime().sync()
+        self.clear_fronted()
+        self._traced_records = impl.get_runtime(
+        ).prog.get_kernel_profiler_records()
+
+    def count_results(self):
+        """count statistical results
+        kernels with the same name will be counted in a StatisticalResult
+        """
+        for record in self._traced_records:
+            if self._statistical_results.get(record.name) == None:
+                self._statistical_results[record.name] = StatisticalResult(
+                    record.name)
+            self._statistical_results[record.name].insert_record(
+                record.kernel_time)
+            self._total_time_ms += record.kernel_time
+        self._statistical_results = {
+            k: v
+            for k, v in sorted(self._statistical_results.items(),
+                               key=lambda item: item[1],
+                               reverse=True)
+        }
+
+    def print_info(self, mode=count):
+        '''`count` mode : print the statistical results (min,max,avg time) of Taichi kernels on devices.
+        `trace` mode : print records of launched Taichi kernels
+        Default print mode is `count` mode
+        '''
+        self.update_records()  # trace records
+        self.count_results()  # statistical results
+        #count mode (default) : print statistical results of all kernel
+        if mode == self.count:
+            print(f"====================================="
```

Comment:
nit: fstring supports multiple lines so these prints can be simplified to only a few ;)

Reply:
got it this time :sweat_smile: 
```
            print(f"{patition_line('=',73)}"\
                  f"{_ti_core.arch_name(ti.cfg.arch).upper()} Profiler(count)\n"\
                  f"{patition_line('=',73)}"\
                  f"[      %     total   count |"\
                  f"      min       avg       max   ]"\
                  f" Kernel name\n")
```